### PR TITLE
Improved zip API

### PR DESCRIPTION
**Pull request by dennisvennink**
This PR contains the implementation, documentation, tests and benchmarks for the *Improved `zip` API* [proposal](https://github.com/apple/swift-evolution/pull/941).

### CODE REVIEW

File: stdlib/public/core/Zip.swift [hunk 1]
```diff
@@ -136,13 +147,140 @@ extension Zip2Sequence.Iterator: IteratorProtocol {
 }
 
 extension Zip2Sequence: Sequence {
-  public typealias Element = (Sequence1.Element, Sequence2.Element)
```
Comment (airspeedswift) [inline stdlib/public/core/Zip.swift]: This is intentionally explicit, please don't remove it.

File: stdlib/public/core/Zip.swift [hunk 2]
```diff
@@ -136,13 +147,140 @@ extension Zip2Sequence.Iterator: IteratorProtocol {
...
+
+    let element1 = _baseStream1.next()
+    let element2 = _baseStream2.next()
+
+    if element1 == nil && element2 == nil {
+      _reachedEnd = true
+      return nil
+    }
+
+    return (element1, element2)
```
Comment (airspeedswift) [inline stdlib/public/core/Zip.swift]: Using `} else {` for the return is probably cleaner here.

File: stdlib/public/core/Zip.swift [hunk 3]
```diff
@@ -136,13 +147,140 @@ extension Zip2Sequence.Iterator: IteratorProtocol {
...
+  ///
+  /// - Note: To create an instance, use `makeIterator()`.
+  @_fixed_layout // FIXME(sil-serialize-all)
+  public struct Iterator {
+    @usableFromInline // FIXME(sil-serialize-all)
+    internal var _baseStream1: Sequence1.Iterator
+    @usableFromInline // FIXME(sil-serialize-all)
+    internal var _baseStream2: Sequence2.Iterator
+    @usableFromInline // FIXME(sil-serialize-all)
+    internal var _reachedEnd: Bool = false
```
Comment (airspeedswift) [inline stdlib/public/core/Zip.swift]: Unlike with regular `zip`, I don't think it's necessary to track this. You can just rely on the two iterators.

**Comment by airspeedswift** [comment]
Can you add a benchmark as well? Thanks!

File: stdlib/public/core/Zip.swift (hunk 2 above)
Comment (dennisvennink) [reply stdlib/public/core/Zip.swift]: `return element1 == nil && element2 == nil ? nil : (element1, element2)` is also an option.

File: stdlib/public/core/Zip.swift (hunk 1 above)
Comment (dennisvennink) [reply stdlib/public/core/Zip.swift]: 👍

File: stdlib/public/core/Zip.swift (hunk 3 above)
Comment (dennisvennink) [reply stdlib/public/core/Zip.swift]: 👍

**Comment by Lantua** [comment]
In `underestimatdCount` implementation, do we really need check if `result >= 0`?
Many implementations that uses it should already gracefully handle (or simply ignore) this pathological case.

**Comment by dennisvennink** [comment]
> In `underestimatdCount` implementation, do we really need check if `result >= 0`?

While the default implementation `return`s `0`, it is not guaranteed. And because we're using `min(_:_:)` it's probably a good idea to be defensive.

**Comment by dennisvennink** [comment]
> Can you add a benchmark as well? Thanks!

@airspeedswift I've added some `Iterator`-related benchmarks. Do you think we should add others as well? And if so, which?

**Comment by futurejones** [comment]
Can one of the admins verify this patch?

**Comment by shahmishal** [comment]
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))

**Pull request closed without merging**